### PR TITLE
task(p2p): make p2p start on demand

### LINKED ISSUE #418: As a Developer, I need p2p to have delayed initialization, so I can calculate its prerequisites (chain_id) in statemachine
This is needed to make chain_id calculated instead of specifying on node startup.

See #397

### PR DESCRIPTION
**Pull request by akoptelov**
This PR adds possibility to make delayed initialization for p2p part. This is needed for calculating chain_id in the state machine.

fixes #418

### CODE REVIEW

File: node/src/ledger/read/ledger_read_actions.rs [hunk 1]
```diff
@@ -27,7 +27,9 @@ pub enum LedgerReadAction {
 impl redux::EnablingCondition<crate::State> for LedgerReadAction {
     fn is_enabled(&self, state: &crate::State, _time: redux::Timestamp) -> bool {
         match self {
-            LedgerReadAction::FindTodos => state.ledger.read.is_total_cost_under_limit(),
+            LedgerReadAction::FindTodos => {
+                state.p2p.ready().is_some() && state.ledger.read.is_total_cost_under_limit()
```
Comment (binier) [inline node/src/ledger/read/ledger_read_actions.rs:31]: Read requests arent just for p2p. Please revert this change.

File: node/src/state.rs [hunk 2]
```diff
@@ -94,3 +96,124 @@ impl State {
...
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum P2pInitializationError {
+    #[error("p2p is already initialized")]
+    AlreadyInitialized,
+}
+
+#[macro_export]
+macro_rules! p2p_ready {
```
Comment (binier) [inline node/src/state.rs:114]: Can't we have a function with the same logic? I'm worried about growing compilation times.

File: node/testing/src/scenarios/cluster_runner.rs [hunk 3]
```diff
@@ -117,15 +118,15 @@ impl<'a> ClusterRunner<'a> {
         )
     }
 
-    pub fn get_chain_id(&self) -> Option<String> {
+    pub fn get_chain_id(&self) -> Option<ChainId> {
```
Comment (binier) [inline node/testing/src/scenarios/cluster_runner.rs:121]: Do we still need chain id inside cluster?

Comment (binier) [reply node/testing/src/scenarios/cluster_runner.rs:121]: Not sure why we had it in the first place. If in test we are interacting with the ocaml node, we will get the chain_id from there, otherwise chainid didnt matter.

File: node/testing/src/simulator/mod.rs [hunk 4]
```diff
@@ -41,9 +42,9 @@ impl Simulator {
         {
             chain_id
         } else if let Some(node) = runner.ocaml_node(ClusterOcamlNodeId::new_unchecked(0)) {
-            node.chain_id_async().await.unwrap().into_bytes()
+            node.chain_id_async().await.unwrap()
         } else {
-            "<unknown_chain_id>".into()
```
Comment (binier) [inline node/testing/src/simulator/mod.rs:46]: Since chain id calculation will be done inside the node, do we still need this or chain id configuration in rust node at all?

**Review by binier — CHANGES_REQUESTED** [review]

File: node/src/action.rs [hunk 5]
```diff
@@ -53,7 +55,7 @@ impl redux::EnablingCondition<crate::State> for CheckTimeoutsAction {}
 
 #[cfg(feature = "replay")]
 impl redux::EnablingCondition<crate::State> for Action {
-    fn is_enabled(&self, _: &crate::State, _time: redux::Timestamp) -> bool {
-        true
+    fn is_enabled(&self, state: &crate::State, _time: redux::Timestamp) -> bool {
+        !matches!(self, Action::InitializeP2p { .. }) || state.p2p.ready().is_none()
```
[the remark this answers is not in the document]
Comment (akoptelov) [reply node/src/action.rs:59]: @binier `InitializeP2p` shouldn't be triggered twice. How can I express that?

File: node/src/state.rs (hunk 2 above)
Comment (akoptelov) [reply node/src/state.rs:113]: This is mostly for sorting out places where an action need p2p but it is not initialized. It turned into `panic` with proper info easily. Also we should bail out if there's no p2p (`return;`). As to compilation time, such macro rules is a drop in the ocean.

File: node/src/ledger/read/ledger_read_actions.rs (hunk 1 above)
Comment (akoptelov) [reply node/src/ledger/read/ledger_read_actions.rs:31]: Hmm... ok, there was a place where p2p was kind of essential, i.e. we collected ready peers. But for uninitialized p2p it will be just empty list.

File: node/testing/src/simulator/mod.rs (hunk 4 above)
Comment (akoptelov) [reply node/testing/src/simulator/mod.rs:46]: Now that's gone.

File: core/src/chain_id.rs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    tx_pool_max_size: &UnsignedExtendedUInt32StableV1,
+) -> [u8; 32] {
+    let mut hasher = Blake2b256::default();
+    let genesis_timestamp = OffsetDateTime::from_unix_timestamp_nanos(
+        (constants.genesis_state_timestamp.0 .0 .0 * 1000000) as i128,
+    )
+    .unwrap();
+    let time_format =
+        format_description!("[year]-[month]-[day] [hour]:[minute]:[second].[subsecond digits:6]Z");
+    let input = format!(
```
Comment (akoptelov) [inline core/src/chain_id.rs:39]: @Sventimir Here a new string is allocated, that contains concatenated bytes. Can you do this by just passing each part to the `hasher.update()`?

File: core/src/chain_id.rs [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        tx_max_pool_size: &UnsignedExtendedUInt32StableV1,
+    ) -> ChainId {
+        let mut hasher = Blake2b256::default();
+        let constraint_system_hash = constraint_system_digests
+            .iter()
+            .map(|md5| u8_slice_as_hex(md5))
+            .collect::<Vec<String>>()
+            .join("");
+        let genesis_constants_hash = hash_genesis_constants(genesis_constants, tx_max_pool_size);
+        let input = format!(
```
Comment (akoptelov) [inline core/src/chain_id.rs:68]: See my previous comment.

File: core/src/chain_id.rs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+use time::OffsetDateTime;
+
+use std::fmt::{self, Debug, Display, Formatter};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct ChainId([u8; 32]);
+
+fn u8_slice_as_hex(slice: &[u8]) -> String {
```
Comment (akoptelov) [inline core/src/chain_id.rs:15]: Let's use `hex::encode` instead of it.

File: core/src/chain_id.rs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        protocol_transaction_version: u8,
+        protocol_network_version: u8,
+        tx_max_pool_size: &UnsignedExtendedUInt32StableV1,
+    ) -> ChainId {
+        let mut hasher = Blake2b256::default();
+        let constraint_system_hash = constraint_system_digests
+            .iter()
+            .map(|md5| u8_slice_as_hex(md5))
+            .collect::<Vec<String>>()
+            .join("");
```
Comment (akoptelov) [inline core/src/chain_id.rs:66]: Probably something like `Iterator::reduce` will be more appropriate then `collect/join`, without allocating a vector.

File: core/src/chain_id.rs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        let input = format!(
+            "{}{}{}{}{}",
+            genesis_state_hash,
+            constraint_system_hash,
+            u8_slice_as_hex(&genesis_constants_hash),
+            md5_hash(protocol_transaction_version),
+            md5_hash(protocol_network_version)
+        );
+        hasher.update(input.as_bytes());
+        ChainId(hasher.finalize().try_into().unwrap())
```
Comment (akoptelov) [inline core/src/chain_id.rs:77]: There should be a way to make hash type-safe for `[u8; 32]`.

File: core/src/chain_id.rs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    pub fn as_hex(&self) -> String {
```
Comment (akoptelov) [inline core/src/chain_id.rs:80]: As per Rust convention, methods prefixed with `as_` are computationally cheap way to reinterpret the value. Here, `to_hex` will be more appropriate.

File: core/src/chain_id.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        hex::encode(&self.0)
+    }
+
+    pub fn from_hex(s: &str) -> Result<ChainId, hex::FromHexError> {
```
Comment (akoptelov) [inline core/src/chain_id.rs:84]: Let's remove it.

File: core/src/chain_id.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        let h = hex::decode(s)?;
+        let bs = h[..32].try_into().or(Err(hex::FromHexError::InvalidStringLength))?;
+        Ok(ChainId(bs))
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> ChainId {
+        let mut arr = [0u8; 32];
+        arr.copy_from_slice(&bytes[..32]);
```
Comment (akoptelov) [inline core/src/chain_id.rs:92]: It will panic if input is shorter. Let's remove the whole method, or use sized array instead.

**Review by akoptelov — COMMENTED** [review]
@Sventimir can you please take a look to my comments?

File: node/src/action.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tizoc) [reply node/src/action.rs:59]: @akoptelov in this case you need to define a type for this action like in the TODO comment, so that that action is dispatched (and `is_enabled` for that type is checked) instead of `Action::InitializeP2p` directly.

Note that queueing `is_enabled` for `Action` is always called, never directly for the subtype (simplifes a lot of stuff, and having `Action::is_enabled` behave differently from the nested types is also confusing). But somehow during replay it should still return `true` I assume (this here reminded me that I still need to solve that)

File: core/src/chain_id.rs (hunk 6 above)
Comment (tizoc) [reply core/src/chain_id.rs:39]: it will, but what reads better? I don't think these allocations here are something to worry about, this code will be called what, once during startup?

Comment (tizoc) [reply core/src/chain_id.rs:39]: oops, sorry, thought this was about the time format above, in this case it is plain concatenation, so `hasher.update()` will read as well as this version.

Comment (Sventimir) [reply core/src/chain_id.rs:39]: It works like this:
```
    hasher.update(constants.k.to_string().as_bytes());
    hasher.update(constants.slots_per_epoch.to_string().as_bytes());
    hasher.update(constants.slots_per_sub_window.to_string().as_bytes());
    hasher.update(constants.delta.to_string().as_bytes());
    hasher.update(tx_pool_max_size.to_string().as_bytes());
    hasher.update(genesis_timestamp.format(&time_format).unwrap().as_bytes());
```

Comment (Sventimir) [reply core/src/chain_id.rs]: Done.

File: core/src/chain_id.rs (hunk 7 above)
Comment (Sventimir) [reply core/src/chain_id.rs]: Done.

File: core/src/chain_id.rs (hunk 8 above)
Comment (Sventimir) [reply core/src/chain_id.rs]: Done.

File: core/src/chain_id.rs (hunk 10 above)
Comment (Sventimir) [reply core/src/chain_id.rs:65]: Unfortunately, that's not so. `hasher.finalize()` returns `[u8]`, which cannot be safely converted into `[u8; L]`, That's just a shortcoming of the type system, as far as I understand (and I read about it quite a bit today).

File: core/src/chain_id.rs (hunk 12 above)
Comment (Sventimir) [reply core/src/chain_id.rs:72]: But this function is being used down below in the implementation of `Deserialize` trait.

File: core/src/chain_id.rs (hunk 11 above)
Comment (Sventimir) [reply core/src/chain_id.rs:68]: Done.

File: core/src/chain_id.rs (hunk 13 above)
Comment (Sventimir) [reply core/src/chain_id.rs:82]: Okay.

File: node/src/action.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (akoptelov) [reply node/src/action.rs:59]: This action enables p2p, not called in it.

@binier if I just wrap it into another enum, would it help?

[the remark this answers is not in the document]
Comment (akoptelov) [reply node/src/action.rs]: Ok, got that now, with EC for Action. Moved to P2pAction, as you suggested.

File: core/src/chain_id.rs (hunk 10 above)
Comment (akoptelov) [reply core/src/chain_id.rs:65]: OK, I mixed up `multihash` library with e.g. `blake2`, that can use fixed-size outputs. We can use that, later.

**Review by Sventimir — APPROVED** [review]

**Review by binier — APPROVED** [review]
